src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1635,8 +1662,11 @@ void SystemDictionary::methods_do(void f(Method*)) {
   // Walk methods in loaded classes
   MutexLocker ml(ClassLoaderDataGraph_lock);
   ClassLoaderDataGraph::methods_do(f);
-  // Walk method handle intrinsics
-  invoke_method_table()->methods_do(f);
+
+  auto doit = [&] (InvokeMethodKey key, Method* method) {
+    f(method);
+  };
+  _invoke_method_intrinsic_table.iterate_all(doit);
```

Comment:
Hmm, this is problematic. The modification of the old `invoke_method_table()` was protected by `SystemDictionary_lock`, but `invoke_method_table()->methods_do(f)` was not. Maybe it worked because `invoke_method_table()` could be read lock-free?

The trouble is that ResourceHashtable is not lock free. However, we cannot hold `SystemDictionary_lock` here because it might introduce a deadlock with the `f` callback.

I am not sure what can be done here.

Reply:
I'm pretty sure Justin can take the InvokeMethodTable_lock around the methods_do() call, since this lock is the highest ranked lock.  He'll have to reduce the scope of the ClassLoaderDataGraph_lock above it.